modules/csvparser/csv-scanner.c

```diff
@@ -0,0 +1,424 @@
+}
+
+void
+csv_scanner_options_set_delimiters(CSVScannerOptions *options, const gchar *delimiters)
+{
+  g_free(options->delimiters);
+  options->delimiters = g_strdup(delimiters);
+}
+
+void
+csv_scanner_options_set_string_delimiters(CSVScannerOptions *options, GList *string_delimiters)
+{
+  string_list_free(options->string_delimiters);
+  options->string_delimiters = string_delimiters;
+}
+
+void
+csv_scanner_options_set_quotes_start_and_end(CSVScannerOptions *options, const gchar *quotes_start, const gchar *quotes_end)
+{
+  g_free(options->quotes_start);
+  g_free(options->quotes_end);
+  options->quotes_start = g_strdup(quotes_start);
+  options->quotes_end = g_strdup(quotes_end);
+}
+
+void
+csv_scanner_options_set_quotes(CSVScannerOptions *options, const gchar *quotes)
+{
+  csv_scanner_options_set_quotes_start_and_end(options, quotes, quotes);
+}
+
+void
+csv_scanner_options_set_quote_pairs(CSVScannerOptions *options, const gchar *quote_pairs)
+{
+  gint i;
+
+  g_free(options->quotes_start);
+  g_free(options->quotes_end);
+
+  options->quotes_start = g_malloc((strlen(quote_pairs) / 2) + 1);
+  options->quotes_end = g_malloc((strlen(quote_pairs) / 2) + 1);
+
+  for (i = 0; quote_pairs[i] && quote_pairs[i+1]; i += 2)
+    {
+      options->quotes_start[i / 2] = quote_pairs[i];
+      options->quotes_end[i / 2] = quote_pairs[i + 1];
+    }
+  options->quotes_start[i / 2] = 0;
+  options->quotes_end[i / 2] = 0;
+}
+
+
+void
+csv_scanner_options_set_null_value(CSVScannerOptions *options, const gchar *null_value)
+{
+  g_free(options->null_value);
+  options->null_value = null_value && null_value[0] ? g_strdup(null_value) : NULL;
+}
+
+void
+csv_scanner_options_copy(CSVScannerOptions *dst, CSVScannerOptions *src)
+{
+  csv_scanner_options_set_delimiters(dst, src->delimiters);
+  csv_scanner_options_set_quotes_start_and_end(dst, src->quotes_start, src->quotes_end);
+  csv_scanner_options_set_null_value(dst, src->null_value);
+  csv_scanner_options_set_string_delimiters(dst, string_list_clone(src->string_delimiters));
+  csv_scanner_options_set_columns(dst, string_list_clone(src->columns));
+  dst->dialect = src->dialect;
+  dst->flags = src->flags;
+}
+
+void
+csv_scanner_options_clean(CSVScannerOptions *options)
+{
+  g_free(options->quotes_start);
+  g_free(options->quotes_end);
+  g_free(options->null_value);
+  g_free(options->delimiters);
+  string_list_free(options->string_delimiters);
+  string_list_free(options->columns);
+}
+
+/************************************************************************
+ * CSVScanner
+ ************************************************************************/
+
+static gboolean
+_is_whitespace_char(const gchar *str)
+{
+  return (*str == ' ' || *str == '\t');
+}
+
+static void
+_skip_whitespace(const gchar **src)
+{
+  while (_is_whitespace_char(*src))
+    (*src)++;
+}
+
+static void
+_switch_to_next_column(CSVScanner *self)
+{
+  if (!self->current_column && self->src && self->src[0])
```

Comment:
Wouldn't it be better to move these two lines into `csv_scanner_parse_input()`? It's a kind of initialization before every `process()` call.